app/Models/CopyRepository.php

```
@@ -190,10 +197,146 @@ public function createManyForBook(int $bookId, string $base, int $howMany, strin
+                }
+                if ($errno === 1062 && $attempt < $maxAttempts) {
+                    continue;
+                }
+                throw new \RuntimeException('Unable to insert allocated copy: ' . $error);
+            }
+        } finally {
+            $this->releaseInventoryAllocatorLock($lockName);
+        }
+
+        throw new \RuntimeException('Unable to allocate a unique inventory code after concurrent retries.');
+    }
```

PHPMD (2.15.0)

[warning] The method createWithAllocatedInventoryCode() has a Cyclomatic Complexity of 10. The configured cyclomatic complexity threshold is 10. (undefined)

(CyclomaticComplexity)